oshi-core/src/main/java/oshi/util/Memoizer.java

```diff
@@ -0,0 +1,79 @@
+/**
+ * OSHI (https://github.com/oshi/oshi)
+ *
+ * Copyright (c) 2010 - 2019 The OSHI Project Team:
+ * https://github.com/oshi/oshi/graphs/contributors
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package oshi.util;
+
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+/**
+ * A memoized function stores the output corresponding to some set of specific
+ * inputs. Subsequent calls with remembered inputs return the remembered result
+ * rather than recalculating it.
+ */
+public final class Memoizer {
+
+    private Memoizer() {
+    }
+
+    /**
+     * Store the function in a thread-safe map to be computed only once.
+     * 
+     * @param f
+     *            The {@link java.util.function.Function} to memoize
+     * @return A memoized version of the function
+     */
+    public static <I, O> Function<I, O> memoize(Function<I, O> f) {
+        ConcurrentMap<I, O> lookup = new ConcurrentHashMap<>();
+        return input -> lookup.computeIfAbsent(input, f);
+    }
+
+    /**
+     * Store a supplier in a delegate function to be computed only once.
+     * 
+     * @param s
+     *            The {@link java.util.function.Supplier} to memoize
+     * @return A memoized version of the supplier
+     */
+    public static <T> Supplier<T> memoize(Supplier<T> s) {
```

Comment:
I have to be honest, your implementation of `memoize` got me into thinking :) It looks like a clever trick, but it has a problem: there is a data race caused by the lack of happens-before relations between write to `delegate` in `getFirst()` and reads from `delegate` in `get()`. As a result, reads from `delegate` in `get()` are allowed to never observe the result of the write to `delegate` in `getFirst()`, i.e. the method `get()` is allowed to always observe the initial delegate calling `getFirst`, and thus the method `get()` is allowed to always acquire a monitor (requested by `synchronized`), which is a much more expensive operation than a volatile read that we have in DCL.

Interestingly, the same logic may not applicable when talking about accesses to `value` provided that `() -> value` captures `value` with the semantics of `final` fields. However, I did not find in JLS any statements saying this. Will ask the [concurrency-interest](http://altair.cs.oswego.edu/mailman/listinfo/concurrency-interest) list because this is interesting.

Anyway, because of the data race I described above, I would suggest to use a straightforward DCL implementation that acquires a monitor only on pessimistic code path, which happens a finite number of times (if the number of threads calling `get()` is final), so the absolute majority of calls is guaranteed to do only a single volatile read per `get()` invocation. Please refer to the comment for `memoize(Function)` for the example of the code.

Reply:
I copied that code directly from [this StackOverflow answer](https://stackoverflow.com/questions/35331327/does-java-8-have-cached-support-for-suppliers/35335467#35335467) (and have added a comment with the source). The comments discuss the happens-before situation.  Note that `getFirst()` is synchronized so only one thread will enter that, updating the values of `initialized` and `delegate` before exiting.  Other threads which may have been waiting for the synchronized method will skip by the `initiialized` check, and any future threads will see the new delegate and never get to the synchronized method at all.

Reply:
I think the explanation of the code in that thread on SO is not correct for the reasons I mentioned earlier here. The implementation also relies on the assumption that captures of variables in lambda expressions have final fields semantics, which may be true (would be reasonable), but I do not see anything about this in JLS. If this assumption does not hold, then the code is completely incorrect; if the assumption holds, then the code has performance problems because the following claim about the synchronized block: "_may happen at most once for every thread, as afterwards, it has the most recent values_" is technically incorrect.

I expressed all my thoughts with more details in the email to concurrency-interest (you are in Bcc). You can follow the thread `another lazy initialization, now with method reference` here [http://cs.oswego.edu/pipermail/concurrency-interest/2019-August/thread.html](http://cs.oswego.edu/pipermail/concurrency-interest/2019-August/thread.html), or by subscribing to the list here [http://altair.cs.oswego.edu/mailman/listinfo/concurrency-interest](http://altair.cs.oswego.edu/mailman/listinfo/concurrency-interest). Let's see what the replies will be, if any (it's an interesting piece of code, and obviously I can be mistaken).